File: shell/components/gnome-shell-extension/lib/multitasking.js
// SPDX-FileCopyrightText: 2026 ComradeArt
// SPDX-License-Identifier: GPL-3.0-or-later
//
// Snap and workspaces. Companion stays bottom-right and does not cover work.
// Geometry is lib/layout.js. Live Mutter tiling is not claimed.

import {SCREEN_QUESTIONS} from './design/tokens.js';
import {
    NAMED_VIEWPORTS, SNAP_TARGETS, snapLayout,
} from './layout.js';

export {NAMED_VIEWPORTS, SNAP_TARGETS, snapLayout};

export const COMPANION_CORNER = 'bottom-right';

export function buildMultitasking({
    workspaces = [], activeId = '', reducedMotion = false,
    screen = NAMED_VIEWPORTS.fhd, scale = 1, profile = 'skeleton',
} = {}) {
    const layout = snapLayout(screen, {scale, profile});
    const items = (Array.isArray(workspaces) ? workspaces : []).map(item => {
        const rec = item && typeof item === 'object' ? item : {name: item};
        return {
            id: String(rec.id || ''),
            name: String(rec.name || 'Workspace'),
            archived: Boolean(rec.archivedAt),
            active: String(rec.id || '') === String(activeId || ''),
        };
    });
    return {
        kind: 'Multitasking',
        title: 'Workspaces',
        summary: 'Snap windows around Bunny. The figure stays bottom-right and does not cover work.',
        companionCorner: COMPANION_CORNER,
        companionCoversWork: layout.companionCoversWork,
        snapTargets: SNAP_TARGETS.slice(),
        snaps: layout.snaps,
        work: layout.work,
        reserved: layout.reserved,
        workspaces: items,
        questions: SCREEN_QUESTIONS,
        doing: 'Arranging windows',
        bunny: 'In the corner',
        next: 'Snap left or right. Maximize leaves the companion visible.',
        companionRequired: false,
        chatbot: false,
        reducedMotion: Boolean(reducedMotion),
        motionMs: reducedMotion ? 0 : 220,
        liveMutterTiling: false,
        accessibleName: 'Workspaces and snap',
        canFocus: true,
        styleClass: 'bunny-multitasking bunny-sheet',
        viewport: `${screen.width}×${screen.height}`,
        scale,
        profile: layout.profile,
    };
}

export function viewportForName(name) {
    const key = String(name || '').toLowerCase();
    if (key === '1366' || key === 'hd1366' || key === 'laptop')
        return NAMED_VIEWPORTS.hd1366;
    if (key === '4k' || key === 'uhd' || key === '2160')
        return NAMED_VIEWPORTS.uhd;
    return NAMED_VIEWPORTS.fhd;
}

File: shell/services/bunny_shell/multitasking.py
# SPDX-FileCopyrightText: 2026 ComradeArt
# SPDX-License-Identifier: GPL-3.0-or-later
"""Snap / workspaces chrome. Geometry lives in layout.js and is tested under node."""

from __future__ import annotations

from typing import Any, Mapping, Sequence

SNAP_TARGETS = ("left", "right", "maximize", "restore")
COMPANION_CORNER = "bottom-right"
NAMED_VIEWPORTS = {
    "hd1366": {"width": 1366, "height": 768, "name": "1366"},
    "fhd": {"width": 1920, "height": 1080, "name": "1080"},
    "uhd": {"width": 3840, "height": 2160, "name": "4K"},
}


def viewport_for_name(name: str | None) -> dict[str, Any]:
    key = str(name or "").casefold()
    if key in {"1366", "hd1366", "laptop"}:
        return dict(NAMED_VIEWPORTS["hd1366"])
    if key in {"4k", "uhd", "2160"}:
        return dict(NAMED_VIEWPORTS["uhd"])
    return dict(NAMED_VIEWPORTS["fhd"])


def build_multitasking(
    *,
    workspaces: Sequence[Mapping[str, Any]] | None = None,
    active_id: str = "",
    reduced_motion: bool = False,
    screen: Mapping[str, Any] | None = None,
    scale: float = 1,
    profile: str = "skeleton",
) -> dict[str, Any]:
    area = dict(screen or NAMED_VIEWPORTS["fhd"])
    items = []
    for item in workspaces or ():
        rec = dict(item)
        ident = str(rec.get("id") or "")
        items.append({
            "id": ident,
            "name": str(rec.get("name") or "Workspace"),
            "archived": bool(rec.get("archivedAt")),
            "active": ident == str(active_id or ""),
        })
    return {
        "kind": "Multitasking",
        "title": "Workspaces",
        "summary": "Snap windows around Bunny. The figure stays bottom-right and does not cover work.",
        "companionCorner": COMPANION_CORNER,
        "companionCoversWork": False,
        "snapTargets": list(SNAP_TARGETS),
        "workspaces": items,
        "companionRequired": False,
        "chatbot": False,
        "reducedMotion": bool(reduced_motion),
        "motionMs": 0 if reduced_motion else 220,
        "liveMutterTiling": False,
        "accessibleName": "Workspaces and snap",
        "canFocus": True,
        "viewport": f"{area.get('width')}×{area.get('height')}",
        "scale": scale,
        "profile": profile,
        "doing": "Arranging windows",
        "bunny": "In the corner",
        "next": "Snap left or right. Maximize leaves the companion visible.",
    }
